### ISSUE TITLE
Document Schema of Database

### ISSUE THREAD

**Issue by mjalkio** [jsfenfen/990-xml-database#16]
Hi @jsfenfen! Thanks for all the work that you've been doing on IRSx and this repo. I'm trying to assess if this database fits my needs, but I'm unfamiliar with Django so I'm unclear on what this generates. Do you think you could add some basic documentation on the schema generated? Do you feel it would make sense to use to generate tables that will be generated with raw SQL queries?

If you have a running database, the output from:
```
SELECT *
FROM information_schema.columns
```

Would be very helpful to me to take a look at. Thanks!

**Comment by jsfenfen** [comment]
> Do you feel it would make sense to use to generate tables that will be generated with raw SQL queries?

That's how I query it. You gotta use double quotes in postgres on variables with internal capital letters, which is annoying, but I've kept them because it makes cryptic seeming variable names more human readable.

> Do you think you could add some basic documentation on the schema generated?

There's actually a whole reference on the schema at [http://www.irsx.info/](http://www.irsx.info/). The basic idea is each database table is a form part (i.e. schedule A part IV) or a repeating group within a form part (i.e. board of directors from schedule J, where each director gets a row). Here's [990 Part I](http://www.irsx.info/metadata/parts/part_i.html). The "variable table" and "variable name" are the table and column names used. 

> I'm unfamiliar with Django so I'm unclear on what this generates

<b>Question:</b> Is there a different ORM you do use? I initially had a SQLAlchemy version of the generate tables code, but never built out the loading code (and am probably unlikely to do so anytime soon...). My sense is it doesn't matter a ton to folks interested in this repo, since the point here is to have queryable DB columns and the ORM is just an intermediate step, but... 


Since you asked, Django uses what is essentially a database schema in [this models file](https://github.com/jsfenfen/990-xml-database/blob/master/irsdb/return/models.py). It'll probably seem familiar to you, but here's the psql description of Part I of the main form 990, which you can, cross reference in the models file [here](https://github.com/jsfenfen/990-xml-database/blob/master/irsdb/return/models.py#L126). Django also prepends the "app" name to the database table, so instead of just part_i you get return_part_i, as below. Also: every table in the db automatically gets an `ein` and `object_id` column added (and at some point there's an index added to them, but you wanna do this after loading). While I'm at it: Schedule K tables get a `documentId` as the IRS allows multiple schedule K forms to be filed and the documentId var keep straight which one. 


    irs=> \d return_part_i;
                                             Table "public.return_part_i"
             Column          |         Type          |                         Modifiers                          
    -------------------------+-----------------------+------------------------------------------------------------
     id                      | integer               | not null default nextval('return_part_i_id_seq'::regclass)
     object_id               | character varying(31) | 
     ein                     | character varying(15) | 
     ActvtyOrMssnDsc         | text                  | 
     CntrctTrmntnInd         | character varying(1)  | 
     VtngMmbrsGvrnngBdyCnt   | bigint                | 
     VtngMmbrsIndpndntCnt    | bigint                | 
     TtlEmplyCnt             | bigint                | 
     TtlVlntrsCnt            | bigint                | 
     TtlGrssUBIAmt           | bigint                | 
     NtUnrltdBsTxblIncmAmt   | bigint                | 
     PYCntrbtnsGrntsAmt      | bigint                | 
     CYCntrbtnsGrntsAmt      | bigint                | 
     PYPrgrmSrvcRvnAmt       | bigint                | 
     CYPrgrmSrvcRvnAmt       | bigint                | 
     PYInvstmntIncmAmt       | bigint                | 
     CYInvstmntIncmAmt       | bigint                | 
     PYOthrRvnAmt            | bigint                | 
     CYOthrRvnAmt            | bigint                | 
     PYTtlRvnAmt             | bigint                | 
     CYTtlRvnAmt             | bigint                | 
     PYGrntsAndSmlrPdAmt     | bigint                | 
     CYGrntsAndSmlrPdAmt     | bigint                | 
     PYBnftsPdTMmbrsAmt      | bigint                | 
     CYBnftsPdTMmbrsAmt      | bigint                | 
     PYSlrsCmpEmpBnftPdAmt   | bigint                | 
     CYSlrsCmpEmpBnftPdAmt   | bigint                | 
     PYTtlPrfFndrsngExpnsAmt | bigint                | 
     CYTtlPrfFndrsngExpnsAmt | bigint                | 
     CYTtlFndrsngExpnsAmt    | bigint                | 
     PYOthrExpnssAmt         | bigint                | 
     CYOthrExpnssAmt         | bigint                | 
     PYTtlExpnssAmt          | bigint                | 
     CYTtlExpnssAmt          | bigint                | 
     PYRvnsLssExpnssAmt      | bigint                | 
     CYRvnsLssExpnssAmt      | bigint                | 
     TtlAsstsBOYAmt          | bigint                | 
     TtlAsstsEOYAmt          | bigint                | 
     TtlLbltsBOYAmt          | bigint                | 
     TtlLbltsEOYAmt          | bigint                | 
     NtAsstsOrFndBlncsBOYAmt | bigint                | 
     NtAsstsOrFndBlncsEOYAmt | bigint                | 
    Indexes:
        "return_part_i_pkey" PRIMARY KEY, btree (id)
        "xx_return_part_i" btree (ein, object_id)`

**Comment by mjalkio** [comment]
> Question: Is there a different ORM you do use?

No, I'm used to skipping ORMs completely, and just living right in the database :)

Why did you decide to split out individual parts as separate tables? Would it not have made sense to consolidate them?

Also, how are you able to place repeating groups into the correct part? For example, how do you programatically know that `SttsWhrCpyOfRtrnIsFldCd` comes from part VI? I see you've mapped it there on http://www.irsx.info/ but I'm unclear on how you would come to that conclusion using IRSx.

**Comment by jsfenfen** [comment]
> Why did you decide to split out individual parts as separate tables? Would it not have made sense to consolidate them?

I started out with them as one per form, but found this hard to work with and ditched it. 

One challenge is guaranteeing a unique (and hopefully meaningful) column name for all these variables. But the requirement is unique per db table so reducing the table size means there's less crazy names, depending on how you write this.

The thing that made me actually give up, though, was mysql running into crazy per-row limitations, I think innodb has a [hard limit of 1017 columns per table](https://dev.mysql.com/doc/refman/8.0/en/innodb-restrictions.html) but maybe also memory-per-column. (Also: why 1017, that's such a random number?) I remember changing to myisam the less robust storage engine as a result,  but when I did I started hitting crazy loading issues where loads would fail and I'd find the db in a hosed unresponsive state and all the advice was "switch to innodb". Just splitting it into smaller tables solved all that (though tbh I don't personally use mysql on this stuff when given the choice). 

Also: having, like, 1,200 columns in myisam was super annoying, I'm used to constantly referring to the database schema directly and I was going blind, I'm sure better db software would help, but... 

> For example, how do you programatically know that SttsWhrCpyOfRtrnIsFldCd comes from part VI?

All this sorta info comes from the .xsd files and I've tried to distill into the [metadata repo](https://github.com/jsfenfen/990-xml-metadata/) which is really just a bunch of csvs with info about the xpaths and separate files for tables and documentation.  W/r/t to the parent form part I think the short answer is that it was a hack.
